lib/Dialects/LinalgTransform/Transforms/TransformInterpreter.cpp

```diff
@@ -494,139 +495,37 @@ executeLowerToLLVMOp(FuncOp target,
-                                     Operation *target) {
-  Block *applyBlock =
-      &sequence->getParentOfType<transform::ApplyOp>().transforms().front();
-  if (applyBlock->getNumArguments() != 1)
-    return sequence.emitError()
-           << "only single-argument sequence blocks are supported";
-
-  DenseMap<Value, Operation *> operations;
-  operations.try_emplace(applyBlock->getArgument(0), target);
-
-  // Stash the function for the cases where we need to apply function-wise
-  // transforms.
-  FuncOp func = target->getParentOfType<FuncOp>();
-  ModuleOp module = func->getParentOfType<ModuleOp>();
-
-  MLIRContext *ctx = target->getContext();
-  RewritePatternSet patternList(ctx);
-  for (Dialect *dialect : ctx->getLoadedDialects())
-    dialect->getCanonicalizationPatterns(patternList);
-  for (RegisteredOperationName op : ctx->getRegisteredOperations())
-    op.getCanonicalizationPatterns(patternList, ctx);
-  FrozenRewritePatternSet patterns(std::move(patternList));
-
-  for (Operation &transform : sequence.body().front()) {
-    if (failed(executeTransform(&transform, func, operations))) {
-      // TODO: should this be a user-visible error?
-      LLVM_DEBUG(DBGS() << "failed to apply transform: " << transform << "\n");
-      return failure();
-    }
-    LLVM_DEBUG(DBGS() << "successfully applied transform: " << transform
-                      << "\n");
-
-    // TODO: remove entries from `operations` if the value is not used by any of
-    // the remaining transformations. This will allow the operation to be
-    // replaced/erased by cleanups below and is slightly more effective.
+namespace {
+/// The only purpose of this class is to enable creation of PatternRewriter
+/// instances as the base class doesn't have a public constructor.
+class SimplePatternRewriter : public PatternRewriter {
+public:
+  explicit SimplePatternRewriter(MLIRContext *ctx) : PatternRewriter(ctx) {}
+};
+} // namespace
 
-    // Run canonicalization, this is similar to running the canonicalizer pass,
-    // but (a) keeps tracking the value/op mapping and (b) avoids constructing
-    // the pattern set + pass pipeline on every step.
-    Operation *canonicalizationRoot = func;
-    if (isa<transform::LowerToLLVMOp>(transform)) {
-      // We cannot run the canonicalizer at a function level past LLVM
-      // conversion, so check that it is the last one.
-      // TODO: this may go away if we have a better scoping mechanism for large
-      // transformations such as bufferization and lowerings and track functions
-      // across them similarly to how we track linalg operations.
-      assert(&transform == &sequence.body().front().back() &&
-             "expected lowering to llvm to be the last transformation");
-      canonicalizationRoot = module;
-    }
+/// Apply at most one pattern from the given list to each operation nested in
+/// `parent`.
+static void applyPatternsOnce(Operation *parent,
```

Comment:
Found this quite confusing as it is very specific to PDL's way of doing things, this would be better as a lambda hidden within `findTransformTarget`.

Reply:
This is not specific to PDL but to the way PDL is used in the transform dialect. I find that a reasonable thing to assume in a static function of a file within transform dialect. Done.